Should DNS discovery use all IPs in a multi-address DNS name as cluster seeds?

My experience is that DNS discovery fails when 1 node out of 3 is down, and discovery spends all of `maxDiscoverAttempts` trying to get gossip from the node that is down, instead of also considering the other 2 nodes' IPs registered with a multi-address DNS name.

I was able to implement the behaviour I expect like this

```java
# ClusterDiscovery.java:

    void discover(ConnectionState state) {
-       List<InetSocketAddress> candidates = new ArrayList<>(this.seeds);
+       List<InetSocketAddress> candidates = new ArrayList<>();
+
+       if (state.getSettings().isDnsDiscover()) {
+           try {
+               InetSocketAddress dnsSeed = this.seeds.get(0);
+
+               // Resolve cluster DNS name
+               candidates = Arrays.stream(InetAddress.getAllByName(dnsSeed.getHostName()))
+                   .map(addr -> new InetSocketAddress(addr, dnsSeed.getPort()))
+                   .collect(Collectors.toList());
+                  
+           } catch (UnknownHostException e) {
+               throw new UncheckedIOException(e);
+           }
+       } else {
+           candidates = new ArrayList<>(this.seeds);
+       }

        if (candidates.size() > 1) {
```
I'm not sure, however, if you'd prefer to delegate somehow this behaviour to the gRPC client, since it's the gRPC client that currently does the lookup.

```
$ host nodes.dev-xxx
nodes.dev-xxx has address 13.55.106.44
nodes.dev-xxx has address 3.104.208.149
nodes.dev-xxx has address 52.64.13.213
```

Node 13.55.106.44 is down, others are up. The issue is present when DNS resolves with 13.55.106.44 as first item in the list.

```
2024-07-08 10:39:35,963 [DEBUG] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] Start connection attempt (1/3)
2024-07-08 10:39:35,963 [DEBUG] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] Using seed node [nodes.dev-xxx/13.55.106.44:443] for cluster node discovery.
2024-07-08 10:39:38,054 [ERROR] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.941476417s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[buffered_nanos=1944060611, waiting_for_connection]]]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:60)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.941476417s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[buffered_nanos=1944060611, waiting_for_connection]]]
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Exception during the node selection process
2024-07-08 10:39:38,059 [ERROR] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] java.util.concurrent.ExecutionException: com.eventstore.dbclient.NoClusterNodeFoundException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.eventstore.dbclient.ConnectionService.createChannel(ConnectionService.java:130)
	at com.eventstore.dbclient.ConnectionService.process(ConnectionService.java:170)
	at com.eventstore.dbclient.RunWorkItem.accept(RunWorkItem.java:30)
	at com.eventstore.dbclient.ConnectionService.run(ConnectionService.java:46)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: com.eventstore.dbclient.NoClusterNodeFoundException: null
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:79)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Error when running discovery process
2024-07-08 10:39:38,060 [DEBUG] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] Start connection attempt (2/3)
2024-07-08 10:39:38,061 [DEBUG] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] Using seed node [nodes.dev-xxx/13.55.106.44:443] for cluster node discovery.
2024-07-08 10:39:40,071 [ERROR] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] java.util.concurrent.TimeoutException: null
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:60)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Exception during the node selection process
2024-07-08 10:39:40,072 [ERROR] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] java.util.concurrent.ExecutionException: com.eventstore.dbclient.NoClusterNodeFoundException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.eventstore.dbclient.ConnectionService.createChannel(ConnectionService.java:130)
	at com.eventstore.dbclient.ConnectionService.process(ConnectionService.java:170)
	at com.eventstore.dbclient.RunWorkItem.accept(RunWorkItem.java:30)
	at com.eventstore.dbclient.ConnectionService.run(ConnectionService.java:46)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: com.eventstore.dbclient.NoClusterNodeFoundException: null
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:79)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Error when running discovery process
2024-07-08 10:39:40,074 [DEBUG] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] Start connection attempt (3/3)
2024-07-08 10:39:40,074 [DEBUG] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] Using seed node [nodes.dev-xxx/13.55.106.44:443] for cluster node discovery.
2024-07-08 10:39:42,078 [ERROR] [] [] [] [ForkJoinPool.commonPool-worker-1] [com.eventstore.dbclient.ClusterDiscovery] java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.997705374s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[buffered_nanos=1999385336, waiting_for_connection]]]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:60)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.997705374s. Name resolution delay 0.000000000 seconds. [closed=[], open=[[buffered_nanos=1999385336, waiting_for_connection]]]
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Exception during the node selection process
2024-07-08 10:39:42,080 [ERROR] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] java.util.concurrent.ExecutionException: com.eventstore.dbclient.NoClusterNodeFoundException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.eventstore.dbclient.ConnectionService.createChannel(ConnectionService.java:130)
	at com.eventstore.dbclient.ConnectionService.process(ConnectionService.java:170)
	at com.eventstore.dbclient.RunWorkItem.accept(RunWorkItem.java:30)
	at com.eventstore.dbclient.ConnectionService.run(ConnectionService.java:46)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: com.eventstore.dbclient.NoClusterNodeFoundException: null
	at com.eventstore.dbclient.ClusterDiscovery.discover(ClusterDiscovery.java:79)
	at com.eventstore.dbclient.ClusterDiscovery.lambda$run$2(ClusterDiscovery.java:42)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Error when running discovery process
2024-07-08 10:39:42,081 [ERROR] [] [] [] [esdb-client-4ff4b4e0-38a4-4030-a0ef-be30fde11ae6] [com.eventstore.dbclient.ConnectionService] Maximum discovery attempt count reached: 3
```

On second thought, I made a temporary DNS multiaddress A record, which I can share here.

```
$ host es-tmp-1.geodesy.ga.gov.au
es-tmp-1.geodesy.ga.gov.au has address 3.24.251.198
es-tmp-1.geodesy.ga.gov.au has address 3.105.10.86
es-tmp-1.geodesy.ga.gov.au has address 3.107.137.137
```

The IP addresses are random; there is no cluster, it's not needed to demonstrate the issue.

Without the above patch, cluster discovery resolves the DNS, but uses only the first IP returned, not all 3.

```
2024-09-23 06:41:11,119 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (1/3)
2024-09-23 06:41:11,119 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 06:41:14,205 [ERROR] ...
2024-09-23 06:41:14,707 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (2/3)
2024-09-23 06:41:14,708 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 06:41:17,712 [ERROR] ...
2024-09-23 06:41:18,213 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (3/3)
2024-09-23 06:41:18,213 [DEBUG] [c.e.dbclient.ConnectionService] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 06:41:21,215 [ERROR] ...
2024-09-23 06:41:21,716 [ERROR] [c.e.dbclient.ConnectionService] Maximum discovery attempt count reached: 3
```

With the above patch, cluster discovery resolves the DNS, grabs all 3 IPs, and tries them each 3 times.

```
2024-09-23 07:41:51,561 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (1/3)
2024-09-23 07:41:51,562 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 07:41:54,628 [ERROR] ...
2024-09-23 07:41:54,629 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.105.10.86:443] for cluster node discovery.
2024-09-23 07:41:54,743 [ERROR] ...
2024-09-23 07:41:54,744 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.107.137.137:443] for cluster node discovery.
2024-09-23 07:41:57,748 [ERROR] ...
2024-09-23 07:41:58,250 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (2/3)
2024-09-23 07:41:58,251 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.105.10.86:443] for cluster node discovery.
2024-09-23 07:41:58,343 [ERROR] ...
2024-09-23 07:41:58,344 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.107.137.137:443] for cluster node discovery.
2024-09-23 07:42:01,349 [ERROR] ...
2024-09-23 07:42:01,350 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 07:42:04,355 [ERROR] ...
2024-09-23 07:42:04,856 [DEBUG] [c.e.dbclient.ConnectionService] Start connection attempt (3/3)
2024-09-23 07:42:04,857 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.107.137.137:443] for cluster node discovery.
2024-09-23 07:42:07,861 [ERROR] ...
2024-09-23 07:42:07,861 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.105.10.86:443] for cluster node discovery.
2024-09-23 07:42:07,938 [ERROR] ...
2024-09-23 07:42:07,938 [DEBUG] [c.e.dbclient.ClusterDiscovery] Using seed node [es-tmp-1.geodesy.ga.gov.au/3.24.251.198:443] for cluster node discovery.
2024-09-23 07:42:10,940 [ERROR] ...
2024-09-23 07:42:11,442 [ERROR] [c.e.dbclient.ConnectionService] Maximum discovery attempt count reached: 3
```

Please let me know if you can reproduce the issue using `es-tmp-1.geodesy.ga.gov.au`.

--- a/db-client-java/src/main/java/com/eventstore/dbclient/ClientFeatureFlags.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/ClientFeatureFlags.java
@@ -1,0 +1,10 @@
+package com.eventstore.dbclient;
+
+public final class ClientFeatureFlags {
+    /**
+     * Enables direct DNS name resolution, retrieving all IP addresses associated with a given hostname. This
+     * functionality was initially implemented to support the now-deprecated TCP API. It is particularly useful in
+     * scenarios involving clusters, where node discovery is enabled.
+     */
+    public static final String DNS_LOOKUP = "dns-lookup";
+}

--- a/db-client-java/src/main/java/com/eventstore/dbclient/ClusterDiscovery.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/ClusterDiscovery.java
@@ -1,5 +1,9 @@
 package com.eventstore.dbclient;
 
+import com.eventstore.dbclient.resolution.DeferredNodeResolution;
+import com.eventstore.dbclient.resolution.DeprecatedNodeResolution;
+import com.eventstore.dbclient.resolution.FixedSeedsNodeResolution;
+import com.eventstore.dbclient.resolution.NodeResolution;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -13,15 +17,18 @@ import java.util.concurrent.TimeoutException;
 class ClusterDiscovery implements Discovery {
     private static final Logger logger = LoggerFactory.getLogger(ClusterDiscovery.class);
     private final NodeSelector nodeSelector;
-    private final List<InetSocketAddress> seeds;
+    private final NodeResolution resolution;
 
     ClusterDiscovery(EventStoreDBClientSettings settings) {
         this.nodeSelector = new NodeSelector(settings.getNodePreference());
 
         if (settings.isDnsDiscover()) {
-            this.seeds = Collections.singletonList(settings.getHosts()[0]);
+            if (settings.getFeatures().contains(ClientFeatureFlags.DNS_LOOKUP))
+                this.resolution = new DeprecatedNodeResolution(settings.getHosts()[0]);
+            else
+                this.resolution = new DeferredNodeResolution(settings.getHosts()[0]);
         } else {
-            this.seeds = Arrays.asList(settings.getHosts());
+            this.resolution = new FixedSeedsNodeResolution(settings.getHosts());
         }
     }
 
@@ -43,7 +50,7 @@ class ClusterDiscovery implements Discovery {
     }
 
     void discover(ConnectionState state) {
-        List<InetSocketAddress> candidates = new ArrayList<>(this.seeds);
+        List<InetSocketAddress> candidates = resolution.resolve();
 
         if (candidates.size() > 1) {
             Collections.shuffle(candidates);

--- a/db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/ConnectionSettingsBuilder.java
@@ -31,6 +31,7 @@ public class ConnectionSettingsBuilder {
     private Long _defaultDeadline = null;
     private List<ClientInterceptor> _interceptors = new ArrayList<>();
     private String _tlsCaFile = null;
+    private Set<String> _features = new HashSet<>();
 
     ConnectionSettingsBuilder() {}
 
@@ -54,7 +55,8 @@ public class ConnectionSettingsBuilder {
                 _keepAliveInterval,
                 _defaultDeadline,
                 _interceptors,
-                _tlsCaFile);
+                _tlsCaFile,
+                _features);
     }
 
     /**
@@ -216,6 +218,22 @@ public class ConnectionSettingsBuilder {
      */
     public ConnectionSettingsBuilder tlsCaFile(String filepath) {
         this._tlsCaFile = filepath;
+        return this;
+    }
+
+    /**
+     * Add feature flags.
+     */
+    public ConnectionSettingsBuilder features(String... features) {
+        this._features.addAll(Arrays.asList(features));
+        return this;
+    }
+
+    /**
+     * Add feature flag.
+     */
+    public ConnectionSettingsBuilder feature(String feature) {
+        this._features.add(feature);
         return this;
     }
 
@@ -434,6 +452,10 @@ public class ConnectionSettingsBuilder {
                         invalidParamFormat(entry[0], entry[1]);
 
                     userKeyFile = entry[1];
+                    break;
+
+                case "feature":
+                    builder._features.add(value);
                     break;
 
                 default:

--- a/db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBClientSettings.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBClientSettings.java
@@ -4,6 +4,7 @@ import io.grpc.ClientInterceptor;
 
 import java.net.InetSocketAddress;
 import java.util.List;
+import java.util.Set;
 
 /**
  * Gathers all the settings related to a gRPC client with an EventStoreDB database.
@@ -39,6 +40,7 @@ public class EventStoreDBClientSettings {
     private final Long defaultDeadline;
     private final List<ClientInterceptor> interceptors;
     private final String tlsCaFile;
+    private final Set<String> features;
 
     /**
      * If the dns discovery is enabled.
@@ -160,6 +162,11 @@ public class EventStoreDBClientSettings {
         return tlsCaFile;
     }
 
+    /**
+     * Feature flags
+     */
+    public Set<String> getFeatures() { return features; }
+
     EventStoreDBClientSettings(
             boolean dnsDiscover,
             int maxDiscoverAttempts,
@@ -175,7 +182,8 @@ public class EventStoreDBClientSettings {
             long keepAliveInterval,
             Long defaultDeadline,
             List<ClientInterceptor> interceptors,
-            String tlsCaFile
+            String tlsCaFile,
+            Set<String> features
     ) {
         this.dnsDiscover = dnsDiscover;
         this.maxDiscoverAttempts = maxDiscoverAttempts;
@@ -192,6 +200,7 @@ public class EventStoreDBClientSettings {
         this.defaultDeadline = defaultDeadline;
         this.interceptors = interceptors;
         this.tlsCaFile = tlsCaFile;
+        this.features = features;
     }
 
     /**

--- a/db-client-java/src/main/java/com/eventstore/dbclient/resolution/DeferredNodeResolution.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/resolution/DeferredNodeResolution.java
@@ -1,0 +1,18 @@
+package com.eventstore.dbclient.resolution;
+
+import java.net.InetSocketAddress;
+import java.util.Collections;
+import java.util.List;
+
+public class DeferredNodeResolution implements NodeResolution {
+    private final InetSocketAddress address;
+
+    public DeferredNodeResolution(InetSocketAddress address) {
+        this.address = address;
+    }
+
+    @Override
+    public List<InetSocketAddress> resolve() {
+        return Collections.singletonList(address);
+    }
+}

--- a/db-client-java/src/main/java/com/eventstore/dbclient/resolution/DeprecatedNodeResolution.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/resolution/DeprecatedNodeResolution.java
@@ -1,0 +1,27 @@
+package com.eventstore.dbclient.resolution;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DeprecatedNodeResolution implements NodeResolution {
+    private final InetSocketAddress address;
+
+    public DeprecatedNodeResolution(InetSocketAddress address) {
+        this.address = address;
+    }
+
+    @Override
+    public List<InetSocketAddress> resolve() {
+        try {
+            return Arrays.stream(InetAddress.getAllByName(address.getHostName()))
+                    .map(addr -> new InetSocketAddress(addr, address.getPort()))
+                    .collect(Collectors.toList());
+        } catch (UnknownHostException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/db-client-java/src/main/java/com/eventstore/dbclient/resolution/FixedSeedsNodeResolution.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/resolution/FixedSeedsNodeResolution.java
@@ -1,0 +1,18 @@
+package com.eventstore.dbclient.resolution;
+
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.List;
+
+public class FixedSeedsNodeResolution implements NodeResolution {
+    private final InetSocketAddress[] seeds;
+
+    public FixedSeedsNodeResolution(InetSocketAddress[] seeds) {
+        this.seeds = seeds;
+    }
+
+    @Override
+    public List<InetSocketAddress> resolve() {
+        return Arrays.asList(seeds);
+    }
+}

--- a/db-client-java/src/main/java/com/eventstore/dbclient/resolution/NodeResolution.java
+++ b/db-client-java/src/main/java/com/eventstore/dbclient/resolution/NodeResolution.java
@@ -1,0 +1,8 @@
+package com.eventstore.dbclient.resolution;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+public interface NodeResolution {
+    List<InetSocketAddress> resolve();
+}

--- a/db-client-java/src/test/java/com/eventstore/dbclient/misc/ParseValidConnectionStringTests.java
+++ b/db-client-java/src/test/java/com/eventstore/dbclient/misc/ParseValidConnectionStringTests.java
@@ -110,6 +110,14 @@ public class ParseValidConnectionStringTests {
                 Arguments.of(
                         "esdb://127.0.0.1:21573?userCertFile=/path/to/cert&userKeyFile=/path/to/key",
                         "{\"dnsDiscover\":false,\"maxDiscoverAttempts\":3,\"discoveryInterval\":500,\"gossipTimeout\":3000,\"nodePreference\":\"leader\",\"tls\":true,\"tlsVerifyCert\":true,\"throwOnAppendFailure\":true,\"hosts\":[{\"address\":\"127.0.0.1\",\"port\":21573}], \"defaultClientCertificate\": {\"clientCertFile\": \"/path/to/cert\", \"clientKeyFile\": \"/path/to/key\"}}"
+                ),
+                Arguments.of(
+                        "esdb://localhost?feature=foobar",
+                        "{\"dnsDiscover\":false,\"maxDiscoverAttempts\":3,\"discoveryInterval\":500,\"gossipTimeout\":3000,\"nodePreference\":\"leader\",\"tls\":true,\"tlsVerifyCert\":true,\"throwOnAppendFailure\":true,\"hosts\":[{\"address\":\"localhost\",\"port\":2113}], \"features\": \"foobar\"}"
+                ),
+                Arguments.of(
+                        "esdb://localhost?feature=foobar&feature=baz",
+                        "{\"dnsDiscover\":false,\"maxDiscoverAttempts\":3,\"discoveryInterval\":500,\"gossipTimeout\":3000,\"nodePreference\":\"leader\",\"tls\":true,\"tlsVerifyCert\":true,\"throwOnAppendFailure\":true,\"hosts\":[{\"address\":\"localhost\",\"port\":2113}], \"features\": [\"foobar\", \"baz\"]}"
                 )
         );
     }
@@ -214,6 +222,15 @@ public class ParseValidConnectionStringTests {
         tree.get("hosts").elements().forEachRemaining((host) -> {
             builder.addHost(new InetSocketAddress(host.get("address").asText(), host.get("port").asInt()));
         });
+
+        if (tree.get("features") != null) {
+           JsonNode features = tree.get("features");
+
+           if (features.isArray())
+               features.elements().forEachRemaining(feature -> builder.feature(feature.asText()));
+           else
+               builder.feature(features.asText());
+        }
 
         return builder.buildConnectionSettings();
     }

